### PR TITLE
Fix async hierarchy export

### PR DESCRIPTION
**Pull request by henning-gerhardt**
Fixes: #5634 

@BartChris : can you please test this changes for your needs even this pull request is still in draft?

### CODE REVIEW

**Comment by BartChris** [comment]
I also had a look at the problem right now and i came to the same conclusions as you did. I think we also have to remove the check for a succesful export here:

https://github.com/kitodo/kitodo-production/blob/master/Kitodo/src/main/java/org/kitodo/export/ExportDms.java#L211-L219

Otherwise the export of children might stop.

File: Kitodo/src/main/java/org/kitodo/production/helper/tasks/ExportDmsTask.java [hunk 1]
```diff
@@ -67,14 +73,25 @@ private ExportDmsTask(ExportDmsTask source) {
...
     @Override
     public void run() {
+        boolean exportSuccessful;
         try {
-            boolean exportSuccessful = exportDms.startExport(process, this);
-            if (Objects.nonNull(exportDms.getWorkflowTask()) && exportSuccessful) {
+            exportSuccessful = exportDms.startExport(process, this);
+            Task workflowTask = exportDms.getWorkflowTask();
+            if (Objects.nonNull(workflowTask)
+                    && workflowTask.getProcess().getId().equals(process.getId())
```
[the remark this answers is not in the document]
Comment (henning-gerhardt) [reply Kitodo/src/main/java/org/kitodo/production/helper/tasks/ExportDmsTask.java:81]: Yes, this was difficult to discover. Exporting the child process was failing, exporting the parent process was successful but the export task of the child was set to done. Maybe I should add an explanation above?

[the remark this answers is not in the document]
Comment (BartChris) [reply Kitodo/src/main/java/org/kitodo/production/helper/tasks/ExportDmsTask.java:81]: Yes, i think it is good, if we add a explanation here.

**Comment by henning-gerhardt** [comment]
> I also had a look at the problem right now and i came to the same conclusions as you did. Additionally, i think we also have to remove the check for a succesful export here:
> 
> https://github.com/kitodo/kitodo-production/blob/master/Kitodo/src/main/java/org/kitodo/export/ExportDms.java#L211-L219
> 
> Otherwise the export of children might stop.

This is a good question. I moved the code of the method only to a method to shrink it on the original method. But I don't know why this is executed. Maybe @matthias-ronge did this know?

**Comment by BartChris** [comment]
This was introduced here
https://github.com/kitodo/kitodo-production/pull/4888

**Comment by henning-gerhardt** [comment]
Question is - which I can not really answer: should all children be exported even if at least the export of one children was failing or not?

**Comment by BartChris** [comment]
I think i would export them all. If one of the exports fails, you can see in the log that an export failed. And if the logic for exporting the parent is executed again after the child is fixed, only this child will be exported again since it is the only one which does not have the isExported flag set to true.

**Comment by henning-gerhardt** [comment]
My edit in the last comment was not right: in this method all children processes get exported which where not already exported. I don't know when this logic is used and which behaviour is expected. So I don't want to change it without more knowledge.

File: Kitodo/src/main/java/org/kitodo/production/helper/tasks/ExportDmsTask.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (henning-gerhardt) [reply Kitodo/src/main/java/org/kitodo/production/helper/tasks/ExportDmsTask.java:82]: I added an explanation.

**Comment by BartChris** [comment]
I put the original issue description for reference https://github.com/kitodo/kitodo-production/issues/4806 : 

`Currently, the export of hierarchies is initiated from the lower level. Superordinate processes are also exported and, during the export of the superordinate processes, the links mentioned below for processes that have not yet been exported are removed from these. An export of the parent would only export these with the removed links.

Goal: When exporting a parent process, all subordinate processes whose workflow has ended but they have not yet been exported [a scenario where the workflow has no export task] are exported first. Then, the superordinate order is exported (once is sufficient) with all the links contained.`

You are right @matthias-ronge should comment here. My intuition would be: Export all child processes and do not stop the export of other children if a single child fails. This child would not have the attribute "isExported" set to true, so it should not appear as reference in the exported parent.

File: Kitodo/src/main/java/org/kitodo/production/helper/tasks/ExportDmsTask.java [hunk 2]
```diff
@@ -67,14 +73,26 @@ private ExportDmsTask(ExportDmsTask source) {
      */
     @Override
     public void run() {
+        boolean exportSuccessful;
         try {
-            boolean exportSuccessful = exportDms.startExport(process, this);
-            if (Objects.nonNull(exportDms.getWorkflowTask()) && exportSuccessful) {
+            exportSuccessful = exportDms.startExport(process, this);
+            Task workflowTask = exportDms.getWorkflowTask();
+            // close only the task of the original process and not of other processes (parent proces, other child processes)
```
Comment (markusweigelt) [inline Kitodo/src/main/java/org/kitodo/production/helper/tasks/ExportDmsTask.java:80]:
```suggestion
            // close only the task of the original process and not of other processes (parent process, other child processes)
```

**Comment by henning-gerhardt** [comment]
I tested this changes so far as I can but there are maybe some circumstances which I can not test or not aware of it.

**Comment by solth** [comment]
@matthias-ronge & @BartChris can you approve this pull request if you do not have any more change requests?

**Comment by BartChris** [comment]
Taken into account what @matthias-ronge and i said above, i think we have to remove the boolean checks here. In case of an async export, triggering the export on the child processes will return false and export of child processes would stop. 

https://github.com/kitodo/kitodo-production/blob/1109ddda68c9060ce9f18b53f579f8a0f630f96d/Kitodo/src/main/java/org/kitodo/export/ExportDms.java#L210-220

**Comment by henning-gerhardt** [comment]
> Taken into account what @matthias-ronge and i said above, i think we have to remove the boolean checks here. In case of an async export, triggering the export on the child processes will return false immediately and export of child processes would stop.
> 
> https://github.com/kitodo/kitodo-production/blob/1109ddda68c9060ce9f18b53f579f8a0f630f96d/Kitodo/src/main/java/org/kitodo/export/ExportDms.java#L210-220
> 
> I think the first check definitely has to be removed
> 
> ```java
>   private boolean exportCompletedChildren(List<Process> children) throws DataException {
>         for (Process child:children) {
>             if (ProcessConverter.getCombinedProgressAsString(child, false).equals(COMPLETED) && !child.isExported()) {
>                 startExport(child)) 
>             }
>         }
>         return true;
>     }
> ```

I don't see why this change (even the method signature should be changed as the return value is in any case `true` and so you don't need this return value after all) is needed to fix the original issue. In my opinion this is an additional useful change but not relating to the issue.

**Comment by BartChris** [comment]
You are right that removing the boolean check is not necessary to fix the issue here. But it has to be addressed  otherwise exporting children of a parent all will not work in asnyc mode. But we can do that in another issue.

**Review by BartChris — APPROVED** [review]

**Comment by henning-gerhardt** [comment]
@matthias-ronge : I tried your scenario while development many times and I never run in your mentioned error. Can you please provide more details (in which state was all 3 processes, was one or more processes already exported, how did you do the export, ....)  if you can reproduce this?

**Pull request merged**